tensorflow/contrib/avro/ops/parse_avro_record.cc

```diff
@@ -0,0 +1,2058 @@
+                             dense_buffers[i_dense], &values);
+          FillInFromValues<string>(n_elements, n_elements_per_batch,
+                                   default_value,
+                                   dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        case DT_DOUBLE: {
+          CopyVarLen<double>(n_elements, n_elements_per_batch,
+                             dense_buffers[i_dense], &values);
+          FillInFromValues<double>(n_elements, n_elements_per_batch,
+                                   default_value,
+                                   dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        case DT_FLOAT: {
+          CopyVarLen<float>(n_elements, n_elements_per_batch,
+                            dense_buffers[i_dense], &values);
+          FillInFromValues<float>(n_elements, n_elements_per_batch,
+                                  default_value,
+                                  dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        case DT_INT64: {
+          CopyVarLen<int64>(n_elements, n_elements_per_batch,
+                            dense_buffers[i_dense], &values);
+          FillInFromValues<int64>(n_elements, n_elements_per_batch,
+                                  default_value,
+                                  dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        case DT_INT32: {
+          CopyVarLen<int32>(n_elements, n_elements_per_batch,
+                            dense_buffers[i_dense], &values);
+          FillInFromValues<int32>(n_elements, n_elements_per_batch,
+                                  default_value,
+                                  dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        case DT_BOOL: {
+          CopyVarLen<bool>(n_elements, n_elements_per_batch,
+                           dense_buffers[i_dense], &values);
+          FillInFromValues<bool>(n_elements, n_elements_per_batch,
+                                 default_value,
+                                 dense_buffers[i_dense].end_indices, &values);
+          break;
+        }
+        default:
+          CHECK(false) << "Should not happen.";  /// Error when avro type does
+                                                 /// not match with tf type
+      }
+    }
+
+    return Status::OK();
+  }
+
+  // Parses a string into avro fields. ASSUMES that the vector for 'avro_fields'
+  // is empty!
+  //
+  // 'avro_fields' An empty vector where we add the avro_fields.
+  //
+  // 'str' Parse this string into Avro types.
+  //
+  static Status StringToAvroField(const string& str,
+                                  std::vector<AvroField*>* avro_fields_ptr) {
+    // Split into tokens using the separator
+    std::vector<AvroField*>& avro_fields = *avro_fields_ptr;
+
+    std::vector<string> incomplete_tokens = str_util::Split(str, ".");
```

Comment:
While trying out the Avro reader, I noticed that it doesn't seem to handle map keys with periods in them properly. For example, when I had a map key whose name was `Resolved.COMPANY_SIZE:companySize`, I got the following error:

```
2018-08-14 15:06:19.846369: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at parse_avro_record.cc:902 : Invalid argument: Could not find name 'globalFeatures[name=Resolved'.
```

Could you fix this and also add a test for this case?

Reply:
From your error message "globalFeatures[name=Resolved" I see that you are using filters.  Note, that filters conceptually allow you to address any datum (nested...) inside your avro type. For this reason, the names inside the filtering expression must follow the restrictions here: https://avro.apache.org/docs/1.7.7/spec.html#Names.  Sidenote: Filters are an experimental feature. At the moment not all nestings are supported.

Reply:
I think there might be a misunderstanding here. Avro map keys are strings, and the strings can contain any character, including periods. So if you have a feature map

```
features: [{
  'foo.bar': 'hello',
  'value': ...
}]
```

you should still be able to filter on `foo.bar` using `features[foo.bar=hello].value`.

I think the string splitting done here just needs to be enhanced to ignore `.`s inside `[...]`. `features[foo.bar=hello].value` should get split into `features[foo.bar=hello]` and `value`.

Reply:
It's true that you can put any string as the key in a map. 
Filters in maps are not supported.
Key access in maps is supported. 
The syntax for that is myMap[myKey] and myKey should be anything -- I'll change that.

Filter in arrays are supported to a limited extent (only on attributes within the same record that is in the array).
The syntax is myArray[attributeName=filterValue].anotherAttributeName to get all another attributes where the attribute name is equal to filter value.

In the latter case attributeName and filterValue shall adhere to the spec here https://avro.apache.org/docs/1.7.7/spec.html#Names to allow for nesting in later implementations of filters.